### PR TITLE
Add ability to create independent store configurations for different lists of same resource

### LINKED ISSUE #8038: Two lists for the same resource share the same store for their state
**What you were expecting:**

I have 2 lists for the same resource. One list is accessible from the main menu and shows all datasets with a rows per page set to 25. The second list for the same resource is part of a tabform in an edit dialog of another resource and shows related datasets. Here we have rows per page of 10 and different search fields than in the first list.

Both lists should should be completely separated and the state of search fields, order and rows per page in the store should be stored for both instances separately.

**What happened instead:**

When I search for data in the second list or change the sort order there, an entry is written in the store with the state of search fields, order and pagination. After that I open the first list from the main menu. Here we see now 10 rows per page instead of 25 and data is filtered by criteria that we entered in the second list. Unfortunately we cannot change the filter because we have different search fields.

The state is stored with the key `RaStore.resource_name.listParams` and is used for both lists, even they are configured differently.

**Environment**

* React-admin version: 4.2.5

### PR DESCRIPTION
**Pull request by septentrion-730n**
Fix #8038

- [x] Add `customStoreKey` to `useListController` and `useListParams`  
- [ ] Unit tests 
- [x] Documentation

### CODE REVIEW

File: docs/useListController.md [hunk 1]
```diff
@@ -33,6 +33,7 @@ const MyList = () => {
 
 `useListController` expects one object as parameter. All keys are optional.
 
+* [`customStoreKey`](#customStoreKey): additional key used to differenciate the list from another of the same resource
```
Comment (fzaninotto) [inline docs/useListController.md:36]: Everything is "custom". I'd simply call it `storeKey`.

File: docs/useListController.md [hunk 2]
```diff
@@ -54,6 +55,7 @@ import {
 } from 'react-admin';
 
 const MyList = ({
+    customStoreKey = '',
```
Comment (fzaninotto) [inline docs/useListController.md:58]: don't add that to the general example, it brings confusion.

File: docs/useListController.md [hunk 3]
```diff
@@ -85,6 +87,10 @@ const MyList = ({
 };
 ```
 
+## `customStoreKey`
+
+To display multiple lists of the same resource and keep distinct store states for each of them (filters, pagination, selection), specify unique keys with the `customStoreKey` property.
+
```
Comment (fzaninotto) [inline docs/useListController.md:93]: there, you're missing an example.

File: packages/ra-core/src/controller/list/useListParams.ts [hunk 4]
```diff
@@ -77,17 +77,18 @@ export interface ListParams {
...
+    filterDefaultValues,
+    perPage = 10,
+    resource,
+    sort = defaultSort,
 }: ListParamsOptions): [Parameters, Modifiers] => {
     const location = useLocation();
     const navigate = useNavigate();
     const [localParams, setLocalParams] = useState(defaultParams);
-    const storeKey = `${resource}.listParams`;
+    const storeKey = `${resource}${customStoreKey}.listParams`;
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/list/useListParams.ts:91]: No, storeKey should be the provided storeKey, or, if not provided, `${resource}.listParams`

**Review by fzaninotto — CHANGES_REQUESTED** [review]

File: docs/useListController.md (hunk 2 above)
Comment (septentrion-730n) [reply docs/useListController.md:58]: Aren't they the default values?

**Pull request closed without merging**